File: packages/ios_platform_images/ios/Classes/UIImage+ios_platform_images.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import <Flutter/Flutter.h>
#import "UIImage+ios_platform_images.h"

@implementation UIImage (ios_platform_images)
+ (UIImage *)flutterImageWithName:(NSString *)name {
  NSString *filename = [name lastPathComponent];
  NSString *path = [name stringByDeletingLastPathComponent];
  for (int screenScale = [UIScreen mainScreen].scale; screenScale > 1; --screenScale) {
    NSString *key = [FlutterDartProject
        lookupKeyForAsset:[NSString stringWithFormat:@"%@/%d.0x/%@", path, screenScale, filename]];
    UIImage *image = [UIImage imageNamed:key
                                inBundle:[NSBundle mainBundle]
           compatibleWithTraitCollection:nil];
    if (image) {
      return image;
    }
  }
  NSString *key = [FlutterDartProject lookupKeyForAsset:name];
  return [UIImage imageNamed:key inBundle:[NSBundle mainBundle] compatibleWithTraitCollection:nil];
}
@end

File: packages/ios_platform_images/ios/Classes/UIImage+ios_platform_images.swift
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import UIKit
import Flutter

public extension UIImage {
  /// Loads a UIImage from the embedded Flutter project's assets.
  ///
  /// This method loads the Flutter asset that is appropriate for the current
  /// screen.  If you are on a 2x retina device where usually `UIImage` would be
  /// loading `@2x` assets, it will attempt to load the `2.0x` variant.  It will
  /// load the standard image if it can't find the `2.0x` variant.
  ///
  /// For example, if your Flutter project's `pubspec.yaml` lists "assets/foo.png"
  /// and "assets/2.0x/foo.png", calling
  /// `[UIImage flutterImageWithName:@"assets/foo.png"]` will load
  /// "assets/2.0x/foo.png".
  ///
  /// See also https://flutter.dev/docs/development/ui/assets-and-images
  ///
  /// Note: We don't yet support images from package dependencies (ex.
  /// `AssetImage('icons/heart.png', package: 'my_icons')`).
  @objc
  static func flutterImage(withName name: String) -> UIImage? {
    let filename = name.components(separatedBy: "/").last
    let path = name.components(separatedBy: "/").dropLast().joined(separator: "/")
    for screenScale in stride(from: UIScreen.main.scale, to: 1, by: -1) {
      let key = FlutterDartProject.lookupKey(forAsset: "\(path)/\(screenScale)0x/\(filename!)")
      if let image = UIImage(named: key, in: Bundle.main, compatibleWith: nil) {
        return image
      }
    }
    let key = FlutterDartProject.lookupKey(forAsset: name)
    return UIImage(named: key, in: Bundle.main, compatibleWith: nil)
  }
}
